Bug 1186124 - Add "Power button locks immediately" option

cr opened the pull request:
As outlined in [bug 1186124](http://bugzil.la/1186124), this adds a prototype of an
option that immediately requires a pass code after pushing the power button regardless
of the pass code timeout setting.

This is a prototype meant for starting the discussion. No tests, no review, lack of
translations, and perhaps not even the best way to go about this.

autolander commented:
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=aeae4393b562c6dce1c868791618377d71a04f48

autolander commented:
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=6552975298afeccb3f590eec2a673b391b69ed77

snowmantw commented on apps/system/js/screen_manager.js:
```diff
@@ -206,6 +206,14 @@
 
         case 'sleep':
           this.turnScreenOff(true, 'powerkey');
+          var lock_enabled = window.lockScreen.enabled;
```
Please note the convention in this file is camel cases. See `this._deviceLightEnabled` above.

snowmantw commented on apps/system/js/screen_manager.js:
```diff
@@ -206,6 +206,14 @@
         case 'sleep':
           this.turnScreenOff(true, 'powerkey');
+          var lock_enabled = window.lockScreen.enabled;
+          var passcode_enabled = window.lockScreen.passCodeEnabled;
```
And it's improper to read another component's status through no the `request` method of System/Service. This would create a tight coupling that hinders us to modify components individually in the future.

snowmantw replied:
You can rather listen the 'sleep' SCM event in `lockscreen.js` to access these properties and do the changes you need in that file. Both `request` and events are for communications among different components, and to decouple them by reducing the direct method calls.

snowmantw commented on apps/system/lockscreen/js/lockscreen.js:
```diff
@@ -754,6 +777,12 @@
     }
   };
 
+  LockScreen.prototype.forcePassCodeUnlock =
```
As I said in Bug 1186100, this should be renamed and changed for a new implementation.

snowmantw commented on apps/system/lockscreen/js/lockscreen.js:
```diff
@@ -64,6 +64,15 @@
+    * Only Settings Listener should change this value to sync with data
+    * in Settings API.
+    */
+    powerButtonLocksEnabled: false,
```
I think it would be better to put your logic into a new LockScreen component under `system/lockscreen/js` and name it as `lockscreen_powerbutton_manager.js` to deal with all these. Although it seems to be a small component, to avoid growing this legacy LockScreen and put each functions in individual components is important. The new component, as a temporary workaround before we could really break all functions into widgets/components of screen lock, can still hold the LockScreen as it's one argument at starting up to get all resources directly. Note this is different from the opinion I left at `screen_manager.js`, which suggest you not access the LockScreen directly, because ScreenManager is totally a different component with different (but coherent) functions from LockScreen, but these sub-components are all related to LockScreen itself.

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=9d2714e86428151b06283121c547568498aba610

cr commented:
@snowmantw, your feedback was extremely helpful, and all oddities of the previous solution (which bothered me, too) clicked nicely into a consistent place.The latest push to this PR should come close to what you outlined.

cr commented:
I will back out the *'lock-immediately'* patch from this PR once it has landed on master, which is a requirement for this feature moving forward. Until then I think it makes more sense to leave it in while review and discussion are still ongoing.

timdream commented on apps/settings/elements/screen_lock.html:
```diff
@@ -31,6 +31,15 @@ <h1 data-l10n-id="screenLock-header"></h1>
           </span>
         </li>
         <li class="passcode-enabled">
+          <p data-l10n-id="powerbutton-behavior">Powerbutton behavior</p>
```
Remove or correct the English string in HTML, as it will be overwritten by l10n property anyway.

cr replied:
I would keep the string, conforming to the convention in this file. Will correct spelling and indent the HTML to avoid long lines.

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=d5356db16a2a65328f969a3ab8a0aea8b9a7fd90

Commit:
Adding "Power button locks immediately" option (bug 1186124)

This adds a lockscreen settings option to immediately require the
pass code (if configured) after pushing the power button, regardless
of the pass code timeout setting.

This allows to have a long pass code timout as fallback for cases
where the user forgets to manually lock, while giving the user a
choice to leave the device unlocked for long periods of time, for
example during development.

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=074fcce5a82915f572cb7707d7b380026cacd665

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=932fbefaecc2c208ee73cfb420704cf19a9309b9

Commit:
Prevent 'lock-immediately' failure state

This adds a check that prevents 'lock-immediately' from throwing
LockScreenWindowManager into an undefined, permanently-locked state
when received while lockscreen disabled.

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=38addeaccee42b784e1dc3181d994095fbd17756

snowmantw commented on apps/system/lockscreen/js/lockscreen_powerbutton_manager.js:
```diff
@@ -0,0 +1,116 @@
+
+  LockScreenPowerbuttonManager.prototype.start =
+    function lspm_start() {
+      window.SettingsListener.observe(
```
Personally I support what Tim said at

https://groups.google.com/forum/#!topic/mozilla.dev.b2g/ZPiHWOSWQR8

So unless I'm moving old code around I would avoid adding new lines using SettingListeners. If it's not necessary, could you change this and to use plain Settings API

snowmantw commented on apps/system/lockscreen/js/lockscreen_powerbutton_manager.js:
```diff
@@ -0,0 +1,116 @@
+   * everything that is not 'passcode-lock' as the default behavior
+   * 'screen-lock'.
+   */
+  LockScreenPowerbuttonManager.prototype.setPowerbuttonLockBehavior =
```
nit: It's a convention to name a function to handle events/changes as `on-` or `handle-`. There are still some differences between a simple setter and a handler.

snowmantw commented on apps/system/lockscreen/js/lockscreen_powerbutton_manager.js:
```diff
@@ -0,0 +1,116 @@
+   * interprets everything that is not true as its default false.
+   */
+  LockScreenPowerbuttonManager.prototype.setLockScreenEnabled =
+    function lspm_setLockScreenEnabled(value) {
```
I've confirmed that the non-boolean value of this setting value is from the very old days, when the mozSettings still only accept serialized value in strings. So if that's the only concerns I think we can save the code and to treat it as boolean directly.

snowmantw replied:
And there is the naming issue as above.

snowmantw commented on apps/system/js/lock_screen_window_manager.js:
```diff
@@ -240,7 +240,15 @@
+          // Ignore when lockscreen is disabled.
+          // Prevents LSWM ending up in a broken, permanently-locked
+          // state when lock-immediately is sent to disabled lockscreen.
+          if (this.states.enabled) {
```
I think for FMD it need to lock it forcibly no matter whether the enable is true? Or a thief can easily bypass that when the settings is still showing on the screen. And the issue of unable to unlock the phone again after exiting FMD mode as you describe here should be fixed in another bug if we didn't fire one.

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f88b0621f58cd63a1f6ad6f511c02a778e0e67c2
